rfcs/80-curve-trait.md

```diff
@@ -0,0 +1,793 @@
+/// but cannot be empty; invalid parameters will result in an error.
+pub fn new(start: f32, end: f32) -> Result<Self, InvalidIntervalError> { //... }
+
+/// Get the start of this interval.
+pub fn start(self) -> f32 { //... }
+
+/// Get the end of this interval.
+pub fn end(self) -> f32 { //... }
+
+/// Get the length of this interval. Note that the result may be infinite (`f32::INFINITY`).
+pub fn length(self) -> f32 { //... }
+
+/// Returns `true` if this interval is finite.
+pub fn is_finite(self) -> bool { //... }
+
+/// Returns `true` if `item` is contained in this interval.
+pub fn contains(self, item: f32) -> bool { //... }
+
+/// Create an [`Interval`] by intersecting this interval with another. Returns an error if the
+/// intersection would be empty (hence an invalid interval).
+pub fn intersect(self, other: Interval) -> Result<Interval, InvalidIntervalError> { //... }
+
+/// Clamp the given `value` to lie within this interval.
+pub fn clamp(self, value: f32) -> f32 { //... }
+
+/// Get the linear map which maps this curve onto the `other` one. Returns an error if either
+/// interval is infinite.
+pub fn linear_map_to(self, other: Self) -> Result<impl Fn(f32) -> f32, InfiniteIntervalError> { //... }
+
+/// Get an iterator over equally-spaced points from this interval in increasing order.
+/// Returns an error if `points` is less than 2 or if the interval is unbounded.
+pub fn spaced_points(
+    self,
+    points: usize,
+) -> Result<impl Iterator<Item = f32>, SpacedPointsError> {
+```
+
+The `Interval` type also implements `TryFrom<RangeInclusive>`, which may be desirable if you want to use
+the `start..=end` syntax. One of the primary benefits of `Interval` (in addition to these methods) is
+that it is `Copy`, so it is easy to take intervals and throw them around. 
+
+### Sampling
+
+The `Curve::sample` method is not intrinsically constrained by the curve's `domain` interval. Instead,
+implementors of `Curve<T>` are free to determine how samples drawn from outside the `domain` will behave.
+However, variants of `sample` (as well as other important methods) use the `domain` explicitly:
+```rust
+/// Sample a point on this curve at the parameter value `t`, returning `None` if the point is
+/// outside of the curve's domain.
+fn sample_checked(&self, t: f32) -> Option<T> { //... }
+
+/// Sample a point on this curve at the parameter value `t`, clamping `t` to lie inside the
+/// domain of the curve.
+fn sample_clamped(&self, t: f32) -> T { //... }
+```
+
+### The Main Curve API
+
+Now, let us turn our attention back to `Curve<T>`, which exposes a functional API similar to that of `Iterator`.
+We will explore its main components one-by-one. The first of those is `map`:
+```rust
+/// Create a new curve by mapping the values of this curve via a function `f`; i.e., if the
+/// sample at time `t` for this curve is `x`, the value at time `t` on the new curve will be
+/// `f(x)`.
+fn map<S>(self, f: impl Fn(T) -> S) -> impl Curve<S>
```

Comment:
Suggestion: I think the trait bound can actually be weakened to `FnOnce`, since this transformation can only be performed a single time.

Reply:
The function is called every time the output curve is sampled, so I don't think it can be `FnOnce`. Theoretically, it could be `FnMut`, but that becomes pretty weird (in a "you really probably shouldn't be doing that" kind of way) since that would be like mutating something captured by a closure every time the curve is sampled.

Reply:
`FnMut` requires mutable access, which would be impossible to call given `sample` does not take a mutable reference, and that would be weird too, to have sampling indicate potential state mutation.

Interior mutability is required, which means either `RefCell` and losing `Send`, or `Arc<Mutex<>>` which would theoretically lock every sample, which has big performance implications.

I think it makes sense here to keep it as `Fn`,  it is being called several times but should have the property that mapping the same sampled point multiple times should give the same result.